yt/frontends/gadget/fields.py

```diff
@@ -65,31 +67,107 @@ def _Metal_density(field, data):
-                ret = data[ptype, "InternalEnergy"] * (gamma - 1) * mu * mp / kb
-                return ret.in_units(self.ds.unit_system["temperature"])
+        This gets used with the magneticum_box2_hr format
+        as defined in the gadget_field_specs in frontends/gadget/definitions.py
+        """
+        metal_names = ["He", "C", "Ca", "O", "N", "Ne", "Mg", "S", "Si", "Fe", "Ej"]
+        for i, metal_name in enumerate(metal_names):
 
-        else:
+            # add the metal fraction fields
+            def _Fraction_wrap(i):
+                def _Fraction(field, data):
+                    return (
+                        data[(ptype, "ElevenMetalMasses")][:, i] / data[(ptype, "Mass")]
+                    )
 
-            def _temperature(field, data):
-                gamma = 5.0 / 3.0
-                ret = data[ptype, "InternalEnergy"] * (gamma - 1) * data.ds.mu * mp / kb
-                return ret.in_units(self.ds.unit_system["temperature"])
+                return _Fraction
+
+            self.add_field(
+                (ptype, metal_name + "_fraction"),
+                sampling_type="particle",
+                function=_Fraction_wrap(i),
+                units="",
+            )
+
+            # add the metal density fields
+            def _Density_wrap(i):
+                def _Metal_density(field, data):
+                    return (
+                        data[(ptype, metal_name + "_fraction")]
+                        * data[(ptype, "density")]
+                    )
+
+                return _Metal_density
+
+            self.add_field(
+                (ptype, metal_name + "_density"),
+                sampling_type="particle",
+                function=_Density_wrap(i),
+                units=self.ds.unit_system["density"],
+            )
+
+        # hydrogen fraction and density
+        def _h_fraction(field, data):
+            ret = data[(ptype, "ElevenMetalMasses")].sum(axis=1) / data[(ptype, "Mass")]
+            return 1.0 - ret
+
+        self.add_field(
+            (ptype, "H_fraction"),
+            sampling_type="particle",
+            function=_h_fraction,
+            units="",
+        )
+
+        def _h_density(field, data):
+            return data[(ptype, "H_fraction")] * data[(ptype, "density")]
 
         self.add_field(
-            (ptype, "Temperature"),
+            (ptype, "H_density"),
             sampling_type="particle",
-            function=_temperature,
-            units=self.ds.unit_system["temperature"],
+            function=_h_density,
+            units=self.ds.unit_system["density"],
         )
+
+    def setup_gas_particle_fields(self, ptype):
+        if (ptype, "Temperature") not in self.ds.field_list:
+
+            if (ptype, "ElectronAbundance") in self.ds.field_list:
+
+                def _temperature(field, data):
+                    # Assume cosmic abundances
+                    x_H = _primordial_mass_fraction["H"]
```

Comment:
Hmm. we need to think about this one a bit more carefully. I didn't change the field definition--I just moved it within a new if-then block. But I see what you mean. I'll try changing it, but I suspect it will break answer tests.

Reply:
@jzuhone pointed me to the [TNG documentation](https://www.tng-project.org/data/docs/faq/#gen5) where they themselves are calculating temperature using this formula, with primordial hydrogen abundances.  So we should mimic that behavior and I retract my point above.